### PR TITLE
Fix issues with syntax errors in policy_json/policy_document

### LINKED ISSUE #3404: iam_policy options 'policy_json' and 'policy_document' give syntax error
##### ISSUE TYPE
- Bug Report
##### COMPONENT NAME

iam_policy
##### ANSIBLE VERSION

```
ansible 2.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```
##### SUMMARY

I'm trying to use either "policy_json" or "policy_document", (as suggested in google - tried with all possible workarounds and methods to pass JSON to it (load from template, file, pass into variable and use from there, escaping double quotes, adding spaces like suggested [here](https://github.com/ansible/ansible-modules-core/pull/3019), and finally simplifying my policy down to 1 line only:    { "Version": "2012-10-17" }      )
but nothing works no matter what.

using -vvv debug I see it whether converting double quotes to single, or adding escape symbols and \n (when it was multiline policy) but always same result: "Syntax errors in policy."

It seems like it doesn't work at all, I can't assign any custom policy using this module (only attaching managed policy works, but not creating inline policy).
##### STEPS TO REPRODUCE

```
attempt 1, using file: 
My policy file content:
{ "Version": "2012-10-17" }

iam_policy:
    iam_type: role
    iam_name: "{{ global.customer }}"
    policy_name: "client-folders"
    state: present
    policy_document: "policy.json"

RESULT:
fatal: [localhost]: FAILED! => {"changed": true, "failed": true, "invocation": {"module_args": {"aws_access_key": null, "aws_secret_key": null, "ec2_url": null, "iam_name": "iamr-ac-ansiblecustomer", "iam_type": "role", "policy_document": "policy.json", "policy_json": null, "policy_name": "iamp-ac-client-folders", "profile": null, "region": null, "security_token": null, "skip_duplicates": true, "state": "present", "validate_certs": true}, "module_name": "iam_policy"}, "msg": "Syntax errors in policy."}

attempt 2 using template:
    policy_json: " '{{ lookup( 'template', 'templates/client-folders.json.j2') }}' "

also doesn't work no matter whats in template... 

attempt 3, put string straight into module:
policy_json: "{\"Version\": \"2012-10-17\"}"
policy_json: ' {"Version":"2012-10-17"} '
policy_json: "{{ some_var_with_policy_json }}"
policy_json: " {{ some_var_with_policy_json }}"
policy_json: "{{ some_var_with_policy_json | to_json }}"

even writing this thing as YAML and passing to  "| to_json"  didn't work  
everything gets turned whether to single quotes or plainly ugly format like:

------------------------------

 "' {\n    \"Version\": \"2012-10-17\",\n    \"Statement\": [\n        {\n            \"Sid\": \"AllowListingOfUserFolder\",\n            \"Action\": [\n                \"s3:ListBucket\"\n            ],\n            \"Effect\": \"Allow\",\n            \"Resource\": [\n  

( I see it from debug  -vvvv )
-------------------------------
```

##### EXPECTED RESULTS

Expected the policy to be assigned
##### ACTUAL RESULTS

```

fatal: [localhost]: FAILED! => {"changed": true, "failed": true, "invocation": {"module_args": {"aws_access_key": null, "aws_secret_key": null, "ec2_url": null, "iam_name": "testing-ansible", "iam_type": "role", "policy_document": null, "policy_json": "' {\n    \"Version\": \"2012-10-17\",\n    \"Statement\": [\n        {\n            \"Sid\": \"AllowListingOfUserFolder\",\n            \"Action\": [\n                \"s3:ListBucket\"\n            ],\n            \"Effect\": \"Allow\",\n            \"Resource\": [\n                \"arn:aws:s3:::mybucket\"\n            ],\n            \"Condition\": {\n                \"StringLike\": {\n                    \"s3:prefix\": [\n                        \"clients/zzz/*\"\n                    ]\n                }\n            }\n        },\n        {\n            \"Sid\": \"AllowAllS3ActionsInUserFolder\",\n            \"Action\": [\n                \"s3:GetObject\"\n            ],\n            \"Effect\": \"Allow\",\n            \"Resource\": [\n                \"arn:aws:s3:::mybucket/clients/zzz/*\"\n            ]\n        }\n    ]\n}'\n", "policy_name": "client-folders", "profile": null, "region": null, "security_token": null, "skip_duplicates": true, "state": "present", "validate_certs": true}, "module_name": "iam_policy"}, "msg": "The policy is not in the valid JSON format."}

```

i hope there will be a solution,
thank you

### PR DESCRIPTION
**Pull request by esc**
Fixes https://github.com/ansible/ansible-modules-core/issues/3404

This is a fix suggested by @robusto:

https://github.com/ansible/ansible-modules-core/issues/3404#issuecomment-231480355

It works for me and for @shaunbrady

### CODE REVIEW

**Comment by gregdek** [comment]
Thanks @esc for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

Core team: please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with 'needs_revision' or merge as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by esc** [comment]
Ping, we are currently having to jump through many hoops to deploy, it's a really trivial fix.

**Comment by tzz** [comment]
Works for me as well, thanks for the great fix @esc

**Comment by sidewinder12s** [comment]
Ping, we've also hit this bug in the module.

File: cloud/amazon/iam_policy.py [hunk 1]
```diff
@@ -283,7 +283,7 @@ def main():
       iam_name=dict(default=None, required=False),
       policy_name=dict(default=None, required=True),
       policy_document=dict(default=None, required=False),
-      policy_json=dict(default=None, required=False),
+      policy_json=dict(type='dict', default=None, required=False),
```
Comment (bcoca) [inline cloud/amazon/iam_policy.py:286]: actually this breaks for those passing in a JSON string, this should be `type=json` in devel and requires some code changes to backport to 2.1

**Comment by esc** [comment]
FWIW: I moved away from using this feature, so I can no longer test if it works or not.

**Comment by ryansb** [comment]
This has been fixed in `devel` and will be part of Ansible 2.2

**Comment by esc** [comment]
ty

**Comment by ryanwalls** [comment]
@ryansb @bcoca @abadger  I still see this issue in 2.2.

**Comment by ryanwalls** [comment]
This:
```
policy_json: "{{ lookup( 'template', 'swf-policy.json.j2') }}"
```

Yields this:
```
"policy_json": "{'Version': '2012-10-17', 'Statement': [{'Action': ['swf:*'], 'Resource': ['arn:aws:swf:*:1234123412341234:/domain/qa'], 'Effect': 'Allow', 'Sid': 'Stmt1486673523000'}]}", 
```

**Comment by ryanwalls** [comment]
The only way to fix it was to manually add the module to my local library and change `type='json'` to `type='dict'`

File: cloud/amazon/iam_policy.py (hunk 1 above)
Comment (ryanwalls) [reply cloud/amazon/iam_policy.py:286]: @bcoca type='json' doesn't work for my json in 2.2.  

My json
```
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "Stmt1486673523000",
      "Effect": "Allow",
      "Action": [
        "swf:*"
      ],
      "Resource": [
        "arn:aws:swf:*:123412341234:/domain/qa"
      ]
    }
  ]
}
```

How I use it in the module:
```
policy_json: "{{ lookup( 'template', 'swf-policy.json.j2') }}"
```

Gives me "Syntax errors" response.  

If I change to `dict` it works.

Comment (ryanwalls) [reply cloud/amazon/iam_policy.py:286]: Also doesn't work if I wrap spaces around it, or use `to_json` or use `convert_data=False`.  Only thing that works is `type='dict'`

Comment (bcoca) [reply cloud/amazon/iam_policy.py:286]: using a template to read a json file actually produces a dict (type discovery will see leading { and convert), but type=json should be able to handle both and produce a json string for the module either way.

Comment (ryanwalls) [reply cloud/amazon/iam_policy.py:286]: @bcoca Thanks.  Yeah, it makes sense.  It just isn't working on my machine with ansible 2.2.

Comment (bcoca) [reply cloud/amazon/iam_policy.py:286]: "just isn't working" is not a helpful error

Comment (ryanwalls) [reply cloud/amazon/iam_policy.py:286]: @bcoca I gave an example here. https://github.com/ansible/ansible-modules-core/pull/4298#issuecomment-278773421.  

The error response is:
```
fatal: [localhost]: FAILED! => {
    "changed": true, 
    "failed": true, 
    "invocation": {
        "module_args": {
            "aws_access_key": null, 
            "aws_secret_key": null, 
            "ec2_url": null, 
            "iam_name": "kubernetes-iam-role-qa", 
            "iam_type": "role", 
            "policy_document": null, 
            "policy_json": "{'Version': '2012-10-17', 'Statement': [{'Action': ['swf:*'], 'Resource': ['arn:aws:swf:*:1234123412341:/domain/qa'], 'Effect': 'Allow', 'Sid': 'Stmt1486673523000'}]}", 
            "policy_name": "qaSwfAllAccessWithinDomain", 
            "profile": null, 
            "region": null, 
            "security_token": null, 
            "skip_duplicates": true, 
            "state": "present", 
            "validate_certs": true
        }, 
        "module_name": "iam_policy"
    }, 
    "msg": "Syntax errors in policy."
}
```

**Comment by bcoca** [comment]
well, the code needs to match the type, in this case the code does not expect json but a dictionary, I would argue the code reading needs to be fixed as the parameter name is misleading otherwise.

File: cloud/amazon/iam_policy.py (hunk 1 above)
Comment (ryansb) [reply cloud/amazon/iam_policy.py:286]: Can you provide the template in the `swf-policy.json.j2` template?

Comment (ryanwalls) [reply cloud/amazon/iam_policy.py:286]: @ryansb Yep, it's here.  https://github.com/ansible/ansible-modules-core/pull/4298#r100413773

**Pull request closed without merging**